### PR TITLE
feat: migrate partners to new backend

### PR DESCRIPTION
**Pull request by ludtkemorgan**
# Pull Request Template

## Issue Overview

This PR addresses #3724 

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

This large PR does the following:
- Re-enables cypress partner tests and get all appropriate tests working again
- Migrates a majority of the typing on partner site from the old backend swagger file to the new one
- Changes all things necessary on the partner site to get the /users page to work
  - Viewing the list of users
  - Adding user
  - Deleting users
  - Editing users
- Changes all things necessary on the partner site to get the /settings page to work
  - Viewing the list of preferences
  - Add new preference
  - Edit existing preference
  - Copy preference
  - Delete preference
- Fix some backend issues found with the integration
  - Listing images and multiselect questions not getting saved properly
  - UserRoles not getting saved properly
  - Wrong user confimation email was getting sent
  - Missing/wrong fields tied to the swagger file

## How Can This Be Tested/Reviewed?

Most of the Partner site now works with the Prisma backend. However in an attempt to make this PR smaller the only part that has been fully tested and integrated with this is all of the `users` and `settings` pages functionality.

After doing the following steps you can log into the the partner site locally and test all of the functionality at `http://localhost:3001/users`.
1. reseed the new backend with `yarn db:setup:staging`
2. Start the new backend and partner sites in separate tabs (`yarn dev` in the appropriate directory)
3. go to http://localhost:3001/ and QA!

This is also deployed to a new prisma partner environment: https://bloom-prisma-partners.netlify.app/

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [ ] I have performed a self-review of my own code
- [ ] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

File: backend_new/prisma/seed-helpers/user-factory.ts [hunk 1]
```diff
@@ -35,6 +36,7 @@ export const userFactory = async (optionalParams?: {
   mfaCodeUpdatedAt: optionalParams?.mfaEnabled ? new Date() : undefined,
   phoneNumber: optionalParams?.phoneNumber || null,
   phoneNumberVerified: optionalParams?.phoneNumberVerified || null,
+  agreedToTermsOfService: optionalParams?.acceptedTerms !== false,
```
Comment (emilyjablonski) [inline backend_new/prisma/seed-helpers/user-factory.ts:39]: Should this be value || null, like above?

File: backend_new/src/dtos/applications/household-member-update.dto.ts [hunk 2]
```diff
@@ -15,12 +15,12 @@ export class HouseholdMemberUpdate extends OmitType(HouseholdMember, [
   @Expose()
   @ValidateNested({ groups: [ValidationsGroupsEnum.default] })
   @Type(() => AddressCreate)
-  @ApiProperty({ type: AddressCreate })
+  @ApiPropertyOptional({ type: AddressCreate })
```
Comment (emilyjablonski) [inline backend_new/src/dtos/applications/household-member-update.dto.ts:18]: Curious about the shift here?

**Review by seanmalbert — CHANGES_REQUESTED** [review]

File: shared-helpers/src/utilities/formKeys.ts [hunk 3]
```diff
@@ -269,7 +269,12 @@ export const listingFeatures = [
   "visual",
 ]
 
-export const applicationLanguageKeys = [Language.en, Language.es, Language.zh, Language.vi]
+export const applicationLanguageKeys = [
```
[the remark this answers is not in the document]
Comment (ludtkemorgan) [reply shared-helpers/src/utilities/formKeys.ts:272]: I just looked up where this is used. It's the dropdown on the paper application form to select the submitted language of the application. We have 5 languages as LanguagesEnum, but for HBA and core we only use 4 of them. I'll switch to the Object.values, but we should do a followup to get the languages from the jurisdiction

File: backend_new/src/dtos/applications/household-member-update.dto.ts (hunk 2 above)
Comment (ludtkemorgan) [reply backend_new/src/dtos/applications/household-member-update.dto.ts:18]: Good point and the old backend also had it required. I'll remove this and address this in the next PR where the application partner flow is tackled

File: backend_new/src/services/user.service.ts [hunk 4]
```diff
@@ -307,9 +307,31 @@ export class UserService {
...
+    if (dto.userRoles && storedUser.userRoles) {
+      if (
+        !(
+          dto.userRoles.isAdmin === storedUser.userRoles.isAdmin &&
+          dto.userRoles.isJurisdictionalAdmin ===
+            storedUser.userRoles.isJurisdictionalAdmin &&
+          dto.userRoles.isPartner === storedUser.userRoles.isPartner
+        )
+      ) {
+        this.prisma.userRoles.update({
```
[the remark this answers is not in the document]
Comment (ludtkemorgan) [reply backend_new/src/services/user.service.ts]: this was unintentional. Fixed!

**Comment by ColinBuyck** [comment]
Still testing but wondering if we can add another jurisdiction to the seed at this point? Feels like there's some jurisdictional admin vs admin user pathways that I could better test with multiple jurisdictions available.

**Comment by ludtkemorgan** [comment]
> Still testing but wondering if we can add another jurisdiction to the seed at this point? Feels like there's some jurisdictional admin vs admin user pathways that I could better test with multiple jurisdictions available.

Great Idea!

**Comment by ludtkemorgan** [comment]
@ColinBuyck a new jurisdiction was added to the seeding

@seanmalbert I think I addressed all of your comments and this is ready to be re-reviewed

**Review by ColinBuyck — COMMENTED** [review]
It seems like the recent changes have surfaced an issue in the email service (potentially multiple jurisdictions?). I'm getting the following error when I invite a new user.

Screenshot 2023-12-04 at 12 25 38 PM

**Comment by ludtkemorgan** [comment]
@ColinBuyck good catch. The failing tests are misleading the error you pointed out causes all subsequent call to the backend to timeout. So the failing test was the emailing step when an Admin user is created. 

I have put up a fix for that and several other use cases that are now exposed with more than one jurisdiction in the system

File: backend_new/src/dtos/applications/household-member-update.dto.ts [hunk 5]
```diff
@@ -15,12 +15,12 @@ export class HouseholdMemberUpdate extends OmitType(HouseholdMember, [
...
   householdMemberAddress: AddressCreate;
 
   @Expose()
   @ValidateNested({ groups: [ValidationsGroupsEnum.default] })
   @Type(() => AddressCreate)
-  @ApiProperty({ type: AddressCreate })
+  @ApiPropertyOptional({ type: AddressCreate })
   householdMemberWorkAddress: AddressCreate;
```
Comment (seanmalbert) [inline backend_new/src/dtos/applications/household-member-update.dto.ts:25]:
```suggestion
  householdMemberWorkAddress?: AddressCreate;
```

**Review by seanmalbert — CHANGES_REQUESTED** [review]

**Comment by seanmalbert** [comment]
Thanks for the updates @ludtkemorgan . I added a few comments to our existing conversations. I'm hoping that @emilyjablonski and @ColinBuyck can provide a more thorough review of the frontend changes.

**Comment by emilyjablonski** [comment]
bug1
bug2

**Comment by ludtkemorgan** [comment]
@emilyjablonski Good catches. Both of these issues have been fixed and I updated some tests. I wish I would have added a second jurisdiction earlier.

**Review by emilyjablonski — APPROVED** [review]

**Review by ColinBuyck — APPROVED** [review]
LGTM 🎂 just one nit on the seeded data, it looks like the Chic Juniperfield preference is assigned to Bloomington listings which should be an impossible state

https://github.com/bloom-housing/bloom/assets/53269332/6bd32e53-15db-4228-87ee-92d8e8f20734

**Pull request merged**